#!/bin/bash -l
#PBS -l walltime=72:00:00,nodes=1:ppn=8,mem=8gb
#PBS -m abe
#PBS -M yang4414@umn.edu
cd $PBS_O_WORKDIR

module load tophat
module load bowtie
module load samtools
module load picard-tools/1.83
module load R
module load cufflinks


newgrp orrharry 

pwd=/home/orrharry/shared/riss/wk12/sample_cleanup
ref_genome_bowtie_index=/panfs/roc/rissdb/genomes/Mus_musculus/mm10_canonical/bowtie2/mm10_canonical
ref_genome_fasta=/panfs/roc/rissdb/genomes/Mus_musculus/mm10_canonical/bowtie2/mm10_canonical.fa

cd /home/orrharry/shared/riss/wk12/tophat_cuffdiff

tophat2 -o FVB_CTTGTA_rep3 -p 4 -r 240 --mate-std-dev 150 --no-novel-juncs -G /panfs/roc/rissdb/igenomes/Mus_musculus/UCSC/mm10/Annotation/Archives/archive-current/Genes/genes.gtf $ref_genome_bowtie_index $pwd/L2_R1_CTTGTA_Orr_009_BLO_02_trimmed_paired.fastq,$pwd/L2_R1_CTTGTA_Orr_009_BLO_02_trimmed_single.fastq,$pwd/L2_R2_CTTGTA_Orr_009_BLO_02_trimmed_single.fastq $pwd/L2_R2_CTTGTA_Orr_009_BLO_02_trimmed_paired.fastq

cd /home/orrharry/shared/riss/wk12/tophat_cuffdiff/FVB_CTTGTA_rep3
java -jar /soft/picard-tools/1.83/CollectInsertSizeMetrics.jar INPUT=accepted_hits.bam OUTPUT=CollectInsertSizeMetrics.out HISTOGRAM_FILE=histogram.pdf 

#REFERENCE_SEQUENCE=$ref_genome_fasta
# Find the mode insert size
# "tail -n +12" :command skips the first 11 lines. It tells tails the first line to print which in this case is 11
#               : The first 11 lines contain header information
#  sort -t  $'\t'  -k2n,2 : Does a numerical sort on the second column. We want to extract the insert size with the highest number of reads (mode)
#  tail -1 : Result of the sort are in increasing order so the last entry has the highest number of reads and hence is the mode insert size.  

tail -n +12 CollectInsertSizeMetrics.out | sort -t  $'\t'  -k2n,2  | tail -1 > insert.txt

cufflinks -p 4 -b $ref_genome_fasta -G /panfs/roc/rissdb/igenomes/Mus_musculus/UCSC/mm10/Annotation/Archives/archive-current/Genes/genes.gtf accepted_hits.bam

